js/loweb.js

```diff
@@ -1,378 +1,302 @@
-        );
-        return {
-          success: (fn) => fn(),
-        };
-      }
-      if (path === '/parse_url') {
-        const providers = getAllProviders();
-        let result;
-        providers.forEach((provider) => {
-          const r = provider.parse_url(request.data.url);
-          if (r !== undefined) {
-            result = r;
-          }
+  clonePlaylist(id, type) {
+    const provider = getProviderByItemId(id);
+    const url = `/playlist?list_id=${id}`;
+    return {
+      success: (fn) => {
+        provider.get_playlist(url).success((data) => {
+          myplaylist.save_myplaylist(type, data);
+          fn();
         });
-        return {
-          success: (fn) => fn({ result }),
-        };
+      },
+    };
+  },
+
+  removeMyPlaylist(id, type) {
+    myplaylist.remove_myplaylist(type, id);
+    return {
+      success: (fn) => fn(),
+    };
+  },
+
+  addMyPlaylist(id, track) {
+    const newPlaylist = myplaylist.add_myplaylist(id, track);
+    return {
+      success: (fn) => fn(newPlaylist),
+    };
+  },
+
+  addPlaylist(id, tracks) {
+    const provider = getProviderByItemId(id);
+    return provider.add_playlist(id, tracks);
+  },
+
+  removeTrackFromMyPlaylist(id, track) {
+    myplaylist.remove_from_myplaylist(id, track);
+    return {
+      success: (fn) => fn(),
+    };
+  },
+
+  removeTrackFromPlaylist(id, track) {
+    const provider = getProviderByItemId(id);
+    return provider.remove_from_playlist(id, track);
+  },
+
+  createMyPlaylist(title, track) {
+    myplaylist.create_myplaylist(title, track);
+    return {
+      success: (fn) => {
+        fn();
+      },
+    };
+  },
+
+  editMyPlaylist(id, title, coverImgUrl) {
+    myplaylist.edit_myplaylist(id, title, coverImgUrl);
+    return {
+      success: (fn) => fn(),
+    };
+  },
+
+  parseURL(url) {
+    const providers = getAllProviders();
+    const result = providers.find((provider) => provider.parse_url(url)).parse_url(url);
+    return {
+      success: (fn) => fn({ result }),
+    };
+  },
+
+  mergePlaylist(source, target) {
+    const tarData = (localStorage.getObject(target)).tracks;
+    const srcData = (localStorage.getObject(source)).tracks;
+    tarData.forEach((tarTrack) => {
+      if (!srcData.find((srcTrack) => srcTrack.id === tarTrack.id)) {
+        myplaylist.add_myplaylist(source, tarTrack);
       }
-      if (path === '/merge_playlist') {
-        const { source, target } = request.data;
+    });
+    return {
+      success: (fn) => fn(),
+    };
+  },
 
-        const tarData = localStorage.getObject(target).tracks;
-        const srcData = localStorage.getObject(source).tracks;
-        tarData.forEach((tarTrack) => {
-          if (!srcData.find((srcTrack) => srcTrack.id === tarTrack.id)) {
-            myplaylist.add_myplaylist(source, tarTrack);
-          }
-        });
-        return {
-          success: (fn) => fn(),
-        };
+  bootstrapTrack(sound, track, playerSuccessCallback, playerFailCallback) {
```

Comment:
auto choose source seems to have some problem. see bug report for more details

Reply:
The tricky usage of Promise makes the `playerFailCallback` is always called.

Reply:
nice, problem solved.